File: 1477-product-of-the-last-k-numbers/1477-product-of-the-last-k-numbers.py
class ProductOfNumbers(object):
  
    def __init__(self):
        self.prefix = [1]
        

    def add(self, num):
        if num > 0:
            self.prefix.append(self.prefix[-1] * num)
        else:
            self.prefix = [1]
        

    def getProduct(self, k):
        n = len(self.prefix)
        return self.prefix[-1] / self.prefix[n - 1 - k] if k < n else 0
        


# Your ProductOfNumbers object will be instantiated and called as such:
# obj = ProductOfNumbers()
# obj.add(num)
# param_2 = obj.getProduct(k)

File: 1477-product-of-the-last-k-numbers/1477-product-of-the-last-k-numbers.cpp
class ProductOfNumbers {
public:
    vector<int> prefix;
    ProductOfNumbers() {
        prefix.push_back(1);
    }
    
    void add(int num) {
        if(num > 0) prefix.push_back(prefix.back() * num); 
        else prefix = {1};
    }
    
    int getProduct(int k) {
        int n = prefix.size();
        return k < n ? prefix.back() / prefix[n - 1 - k] : 0;
    }
};

// k = 3
// 1 4 10 34 55 222
// n-1 = 5, 2


/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * ProductOfNumbers* obj = new ProductOfNumbers();
 * obj->add(num);
 * int param_2 = obj->getProduct(k);
 */